sdk/nodejs/src/hooks/EvalHooksRunner.ts

```diff
@@ -15,31 +19,45 @@ export class EvalHooksRunner {
         key: string,
         defaultValue: T,
         metadata: HookMetadata,
-        resolver: (context: HookContext<T>) => DVCVariable<T>,
+        resolver: (context: HookContext<T>) => VariableWithMetadata<T>,
     ): DVCVariable<T> {
         const context = new HookContext<T>(user, key, defaultValue, metadata)
         const savedHooks = [...this.hooks]
         const reversedHooks = [...savedHooks].reverse()
 
         let beforeContext = context
         let variableDetails: DVCVariable<T>
+        let variableMetadata: VariableMetadata
         try {
             beforeContext = this.runBefore(savedHooks, context)
-            variableDetails = resolver.call(beforeContext)
-            this.runAfter(savedHooks, beforeContext, variableDetails)
+            const { variable, metadata } = resolver.call(beforeContext)
+            variableDetails = variable
+            variableMetadata = metadata
+            this.runAfter(
+                savedHooks,
+                beforeContext,
+                variableDetails,
+                variableMetadata,
+            )
         } catch (error) {
             this.runError(reversedHooks, context, error)
-            this.runFinally(reversedHooks, context, undefined)
+            this.runFinally(reversedHooks, context, undefined, undefined)
             if (
                 error.name === 'BeforeHookError' ||
                 error.name === 'AfterHookError'
             ) {
                 // make sure to return with a variable if before or after hook errors
-                return resolver.call(context)
+                const { variable } = resolver.call(context)
+                return variable
             }
             throw error
         }
-        this.runFinally(reversedHooks, context, variableDetails)
+        this.runFinally(
+            reversedHooks,
+            context,
+            variableDetails,
+            variableMetadata,
+        )
         return variableDetails
```

Comment:
if this is still returning the variable, should the return of this function not be DVCVariable instead of VariableWithMetadata